File: ActiveRealmSample/ActiveRealmSample/Model/Author.m
//
// Created by Masaki Ando on 2019-06-02.
// Copyright (c) 2019 Hituzi Ando. All rights reserved.
//

#import "Author.h"

#import "Article.h"
#import "UserSettings.h"

@interface ActiveRealmAuthor : ARMObject

@property NSString *name;
@property NSNumber <RLMInt> *age;

@end

RLM_ARRAY_TYPE(ActiveRealmAuthor)

@implementation ActiveRealmAuthor

@end

@implementation Author

+ (NSArray<NSString *> *)ignoredProperties {
    return @[ @"shortUID" ];
}

+ (NSDictionary<NSString *, ARMRelationship *> *)definedRelationships {
    return @{
        @"userSettings": [ARMRelationship relationshipWithClass:UserSettings.class
                                                           type:ARMRelationshipTypeHasOne],
        @"articles": [ARMRelationship relationshipWithClass:Article.class
                                                       type:ARMRelationshipTypeHasMany]
    };
}

+ (BOOL)validateBeforeSaving:(id)obj {
    Author *author = obj;

    return author.name.length > 0;
}

- (NSString *)description {
    return self.asJSONString;
}

#pragma mark - property

- (nullable UserSettings *)userSettings {
    return self.relations[@"userSettings"].object;
}

- (NSArray<Article *> *)articles {
    return self.relations[@"articles"].objects;
}

- (NSString *)shortUID {
    return [self.uid componentsSeparatedByString:@"-"].firstObject;
}

#pragma mark - public method

- (NSNumber *)generation:(Author *)obj {
    return @(floorf(self.age.integerValue / 10.f) * 10);
}

- (NSArray<NSString *> *)works:(Author *)obj {
    NSMutableArray *array = [NSMutableArray new];

    for (Article *article in self.articles) {
        [array addObject:article.asJSONString];
    }

    return array;
}

@end

File: ActiveRealmSwiftSample/ActiveRealmSwiftSample/Model/Author.swift
//
// Created by Masaki Ando on 2019-06-04.
// Copyright (c) 2019 Hituzi Ando. All rights reserved.
//

import Foundation

import ActiveRealm

class ActiveRealmAuthor: ARMObject {

    @objc dynamic var name          = ""
    @objc dynamic var age: NSNumber = 0
}

class Author: ARMActiveRealm {

    @objc var name          = ""
    @objc var age: NSNumber = 0

    // Relation properties. These properties are just aliases.

    var userSettings: UserSettings? {
        guard let userSettings = relations["userSettings"]?.object as? UserSettings else { return nil }
        return userSettings
    }

    var articles: [Article] {
        guard let articles = relations["articles"]?.objects as? [Article] else { return [] }
        return articles
    }

    // A property ignored by ActiveRealm.

    @objc var shortID: String {
        return String(uid.split(separator: "-").first!)
    }

    override class func ignoredProperties() -> [String] {
        return ["shortID"]
    }

    override class func definedRelationships() -> [String: ARMRelationship] {
        return [
            "userSettings": ARMRelationship(with: UserSettings.self, type: .hasOne),
            "articles": ARMRelationship(with: Article.self, type: .hasMany)]
    }

    override class func validateBeforeSaving(_ obj: Any) -> Bool {
        let author = obj as! Author

        return !author.name.isEmpty
    }

    override var description: String {
        return asJSONString()
    }
}

extension Author {

    @objc func generation(_ obj: Author) -> NSNumber {
        return NSNumber(integerLiteral: Int(age.doubleValue / 10.0) * 10)
    }

    @objc func works(_ obj: Author) -> [String] {
        return articles.map { article -> String in article.asJSONString() }
    }
}
